### PR TITLE
Python 3.x + Django-CMS 3.3 + Mailchimp API V3.0 ready

### PR DESCRIPTION
**Pull request by matinfo**
Changes:
- fix for python 3.x compatibility
- use python mailchimp3 in place to pyrate to use V3.0 API
- fix app hook for django-cms >=3.3
- fetch_campaigns: now get campaigns in place of lists
- campaign admin view: added list_name, list_id associated to campaign
- keywords: added list_name to filter auto categories when fetch_campaigns
- added *Subscription Advanced* form that include firstname/lastname (FNAME, LNAME) 
- complementary info in README
- remove mc_language plugin form option (no more need, auto detected by Mailchimp)
- template fix for *Campaign Archive* and *Selected Campaign* plugin template to work with django-cms 3.x
- add optional subscription opt-in to plugin forms
- version incremented to 0.4.2

### CODE REVIEW

**Comment by matinfo** [comment]
My test migration/installation, looks good!

File: aldryn_mailchimp/cms_plugins.py [hunk 1]
```diff
@@ -40,6 +40,35 @@ def get_plugin_urls(self):
 plugin_pool.register_plugin(SubscriptionCMSPlugin)
 
 
+class Subscription2CMSPlugin(CMSPluginBase):
```
Comment (Chive) [inline aldryn_mailchimp/cms_plugins.py]: I don't really like the naming, I'd prefer something more descriptive

File: aldryn_mailchimp/cms_plugins.py [hunk 2]
```diff
@@ -40,6 +40,35 @@ def get_plugin_urls(self):
...
+    name = _('Subscription Advanced [email, FNAME and LNAME]')
+    model = SubscriptionPlugin
+    module = _('MailChimp')
+
+    def render(self, context, instance, placeholder):
+        request = context['request']
+        context['form'] = Subscription2PluginForm(
+            initial={'plugin_id': instance.pk,
+                    'redirect_url': request.get_full_path()})
+        return context
```
Comment (Chive) [inline aldryn_mailchimp/cms_plugins.py]: My brain might be lagging behind, but shouldn't we be calling `super` here?

File: aldryn_mailchimp/cms_plugins.py [hunk 3]
```diff
@@ -40,6 +40,35 @@ def get_plugin_urls(self):
...
+
+    def get_subscription_view(self):
+        return Subscription2View.as_view()
+
+    def get_plugin_urls(self):
+        subscription_view = self.get_subscription_view()
+
+        return patterns('',
```
Comment (Chive) [inline aldryn_mailchimp/cms_plugins.py]: please add a linebreak here and then indent normally

File: aldryn_mailchimp/forms.py [hunk 4]
```diff
@@ -5,6 +5,31 @@
...
+    email = forms.EmailField(
+        max_length=254,
+        label=_('Email address'),
+        widget=forms.EmailInput(attrs={'required': True})
+    )
+    plugin_id = forms.CharField(widget=forms.HiddenInput)
+    redirect_url = forms.CharField(widget=forms.HiddenInput)
+
+
+class Subscription2PluginForm(forms.Form):
```
Comment (Chive) [inline aldryn_mailchimp/forms.py]: same as above re naming

File: aldryn_mailchimp/migrations/0004_auto_20170329_1404.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
```
Comment (Chive) [inline aldryn_mailchimp/migrations/0004_auto_20170329_1404.py]: do we need 4 different migrations?

File: aldryn_mailchimp/models.py [hunk 6]
```diff
@@ -13,13 +13,18 @@
...
+    double_optin = models.BooleanField(
+        _('Double Opt-In'), default=True, help_text=_(
+            'If select perform double opt-in.'))
 
     def __str__(self):
-        return unicode(self.list_id)
+        double_optin = 'No'
+        if self.double_optin:
+            double_optin = 'Yes'
+        return u'List ID: %s, Double Opt-In: %s' % (self.list_id, double_optin)
```
Comment (Chive) [inline aldryn_mailchimp/models.py]: please use the new-style string formatting `str.format`

File: aldryn_mailchimp/models.py [hunk 7]
```diff
@@ -105,16 +115,25 @@ class CampaignArchivePlugin(CMSPlugin):
...
 
     def copy_relations(self, old_instance):
         self.categories = old_instance.categories.all()
 
+    def __str__(self):
+        if self.count:
+            count = self.count
+        else:
+            count = 'All'
+        return u'Show: %s, Categories: %s' % (count, ", ".join(str(x) for x in self.categories.all()))
```
Comment (Chive) [inline aldryn_mailchimp/models.py]: same as above, and please translatable.

File: aldryn_mailchimp/models.py [hunk 8]
```diff
@@ -105,16 +115,25 @@ class CampaignArchivePlugin(CMSPlugin):
...
+
 
 class SelectedCampaignsPlugin(CMSPlugin):
     campaigns = models.ManyToManyField(Campaign)
 
     def copy_relations(self, old_instance):
         self.campaigns = old_instance.campaigns.all()
+
+    def __str__(self):
+        return ', '.join(str(x) for x in self.campaigns.all())
```
Comment (Chive) [inline aldryn_mailchimp/models.py:141]: unicode?

File: aldryn_mailchimp/views.py [hunk 9]
```diff
@@ -23,29 +23,82 @@
...
+            SubscriptionPlugin, pk=form.cleaned_data['plugin_id'])
+
+        status = 'subscribed'
+        if plugin.double_optin:
+            status = 'pending'
+
+        try:
+            mclient.lists.members.create(plugin.list_id, {
+                'email_address': form.cleaned_data['email'],
+                'status': 'subscribed',
```
Comment (Chive) [inline aldryn_mailchimp/views.py]: shouldn't this be ``'status': status``?

File: aldryn_mailchimp/views.py [hunk 10]
```diff
@@ -23,29 +23,82 @@
...
+            message = _('Please enter a valid email.')
+            messages.error(self.request, message)
+            response = HttpResponseRedirect(redirect_url)
+        else:
+            # user has tampered with the redirect_url field.
+            response = HttpResponseBadRequest()
+        return response
+
+
+class Subscription2View(FormView):
```
Comment (Chive) [inline aldryn_mailchimp/views.py]: same as above

**Review by Chive — COMMENTED** [review]
Hi @matinfo, thanks a lot for the PR. Good job in decoupling this from pyrate :)

First of all sorry for the delayed response - I completely missed this PR. I've left some minor comments, could you please review those? After that I see no blockers for merging this.

File: aldryn_mailchimp/models.py (hunk 8 above)
Comment (matinfo) [reply aldryn_mailchimp/models.py:141]: Not necessary with Python 3 + decorator @python_2_unicode_compatible:

Django provides a simple way to define __str__() and  __unicode__() methods that work on Python 2 and 3: you must define a __str__() method returning text and to apply the python_2_unicode_compatible() decorator.

On Python 3, the decorator is a no-op. On Python 2, it defines appropriate  __unicode__() and __str__() methods (replacing the original __str__() method in the process).

**Comment by matinfo** [comment]
@Chive In my last changes, I re-worked the mailchimp api v3 error handling in views.py. Plus main of change feedback you requested.

File: aldryn_mailchimp/cms_plugins.py (hunk 2 above)
Comment (matinfo) [reply aldryn_mailchimp/cms_plugins.py]: Was already without super(). Not sure is necessary.

File: aldryn_mailchimp/cms_plugins.py (hunk 1 above)
Comment (matinfo) [reply aldryn_mailchimp/cms_plugins.py]: Done, renamed to SubscriptionExtendedCMSPlugin

File: aldryn_mailchimp/migrations/0004_auto_20170329_1404.py (hunk 5 above)
Comment (matinfo) [reply aldryn_mailchimp/migrations/0004_auto_20170329_1404.py]: Do one new migration.

File: aldryn_mailchimp/forms.py (hunk 4 above)
Comment (matinfo) [reply aldryn_mailchimp/forms.py]: Done

File: aldryn_mailchimp/models.py (hunk 6 above)
Comment (matinfo) [reply aldryn_mailchimp/models.py]: Done

File: aldryn_mailchimp/views.py (hunk 9 above)
Comment (matinfo) [reply aldryn_mailchimp/views.py]: Fixed

File: aldryn_mailchimp/models.py (hunk 7 above)
Comment (matinfo) [reply aldryn_mailchimp/models.py]: Done

File: aldryn_mailchimp/cms_plugins.py (hunk 3 above)
Comment (matinfo) [reply aldryn_mailchimp/cms_plugins.py]: Done

**Comment by koslib** [comment]
Hello guys!
Any updates regarding the progress on this PR?

In any case, I am available for any kind of help needed in order to close this PR, as I am looking forward to use it in production :)

**Comment by matinfo** [comment]
Hi @koslibpro , I will provide a other PR will less fondamental changes, but with the compatibility needs: Python 3, cms 3.3+ and API mailchimp 3. Thx for your patience.

**Comment by koslib** [comment]
@matinfo that's great! Do you have a time estimate about this?

**Comment by matinfo** [comment]
@koslibpro max a week, I hope :-)

**Comment by koslib** [comment]
Awesome @matinfo, I'll keep an eye on this matter then! As I said also earlier, don't hesitate to reach out to me if you need any help!
Cheers!

**Comment by matinfo** [comment]
new PR #22 ready to be accepted by maintainer

**Comment by matinfo** [comment]
This PR no more need, see PR #22

**Comment by koslib** [comment]
thanks @matinfo :)

**Pull request closed without merging**